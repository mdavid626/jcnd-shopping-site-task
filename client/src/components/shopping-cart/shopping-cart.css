.ShoppingCart {
  background-color: #f2f2f2;
  border-radius: 5px;
  padding: var(--space3);
  display: flex;
  flex-direction: column;
  gap: var(--space3);
  min-width: 700px;
  min-height: 400px;
}

.ShoppingCart-header {
  font-size: 22px;
}

.ShoppingCart-content {
  display: flex;
  flex-direction: column;
  flex: 1;
}

.ShoppingCart-close {
  cursor: pointer;
  color: #b3005e;
}

.ShoppingCart-items {
  display: flex;
  flex-direction: column;
  gap: var(--space3);
  overflow-y: scroll;
  max-height: 280px;
  padding: 0 var(--space2) 0 0;
  flex: 1;
}

.ShoppingCart-total {
  align-self: flex-end;
  font-size: 18px;
}

.ShoppingCart-actions {
  display: flex;
  gap: var(--space3);
  align-items: center;
  align-self: flex-end;
}

.ShoppingCart-placeOrder {
  cursor: pointer;
  font-size: 18px;
  background-color: var(--primary-color);
  padding: var(--space2) var(--space3);
  color: #f2f2f2;
  border-radius: 3px;
}
