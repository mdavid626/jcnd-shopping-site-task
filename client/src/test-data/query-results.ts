export const productsQueryResult = {
  data: {
    products: {
      nodes: [
        {
          id: '63ea6485732ff5cf8b364431',
          priceInCents: 9699,
          name: 'Broccoli',
          inStock: 466,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364432',
          priceInCents: 3434,
          name: 'Broccolini',
          inStock: 223,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364433',
          priceInCents: 3692,
          name: 'Brussels Sprouts',
          inStock: 12,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364434',
          priceInCents: 8839,
          name: 'Cabbage',
          inStock: 200,
          __typename: 'Product',
        },
        {
          id: '63ea6485732ff5cf8b364435',
          priceInCents: 6828,
          name: 'Carrots',
          inStock: 151,
          __typename: 'Product',
        },
      ],
      pageInfo: { page: 1, numberOfPages: 15, __typename: 'PageInfo' },
      __typename: 'ProductsQueryResult',
    },
  },
};

export const placeOrderQueryResult1 = {
  data: { placeOrder: '63ede5e031f3fc4012a0951b' },
};
