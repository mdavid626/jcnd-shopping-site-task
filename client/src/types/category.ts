export type Category = 'vegetables' | 'fruits' | 'cheese';
