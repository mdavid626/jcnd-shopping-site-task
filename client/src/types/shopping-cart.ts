export type ShoppingCartItem = {
  productId: string;
  amount: number;
};
