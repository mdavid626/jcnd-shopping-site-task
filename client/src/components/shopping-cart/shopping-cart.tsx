import React from 'react';
import {
  usePlaceOrder,
  useShoppingCart,
} from '../../hooks/shopping-car-thooks/shopping-cart-hooks';
import { ShoppingCartItem } from '../../types/shopping-cart';
import Currency from '../currency/currency';
import ShoppingCartListItem from '../shopping-cart-list-item/shopping-cart-list-item';
import './shopping-cart.css';

const calculateTotal = (shoppingCart: ShoppingCartItem[]) =>
  shoppingCart.reduce(
    (acc, item) => acc + item.product.priceInCents * item.amount,
    0
  );

const ShoppingCart: React.FC<{ onClose: () => void }> = ({ onClose }) => {
  const [shoppingCart] = useShoppingCart();
  const [placeOrder, isPlacingOrder] = usePlaceOrder();
  const handlePlaceOrderClick = async () => {
    await placeOrder();
    onClose();
  };
  return (
    <div className="ShoppingCart">
      <div className="ShoppingCart-header">Shopping Cart</div>
      <div className="ShoppingCart-content">
        {shoppingCart.length > 0 ? (
          <div className="ShoppingCart-items">
            {shoppingCart.map((item) => (
              <ShoppingCartListItem
                key={item.product.id}
                shoppingCartItem={item}
              />
            ))}
          </div>
        ) : (
          <div>Your shopping cart is empty</div>
        )}
      </div>
      {shoppingCart.length > 0 && (
        <div className="ShoppingCart-total" data-testid="ShoppingCart-total">
          Total: <Currency priceInCents={calculateTotal(shoppingCart)} />
        </div>
      )}
      <div className="ShoppingCart-actions">
        <div onClick={() => onClose()} className="ShoppingCart-close">
          close
        </div>
        {shoppingCart.length > 0 && !isPlacingOrder && (
          <div
            onClick={handlePlaceOrderClick}
            className="ShoppingCart-placeOrder"
          >
            Place order
          </div>
        )}
        {isPlacingOrder && (
          <div className="ShoppingCart-placeOrder">Placing order...</div>
        )}
      </div>
    </div>
  );
};

export default ShoppingCart;
