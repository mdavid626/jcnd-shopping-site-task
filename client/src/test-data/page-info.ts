import { PageInfo } from '../types/page-info';

export const testPageInfo1: PageInfo = {
  page: 1,
  numberOfPages: 5,
};
