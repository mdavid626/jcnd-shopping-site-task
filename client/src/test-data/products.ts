import { Product, ProductsQueryResult } from '../types/products';

export const testProduct1: Product = {
  id: '63ea6485732ff5cf8b36442c',
  name: 'Artichoke',
  price: 46.36,
  inStock: 5,
};

export const testProduct2: Product = {
  id: '63ea6485732ff5cf8b36442d',
  name: 'Asparagus',
  price: 25.36,
  inStock: 3,
};

export const testProductQueryResult1: ProductsQueryResult = {
  products: {
    nodes: [testProduct1, testProduct2],
    pageInfo: {
      page: 1,
      numberOfPages: 2,
    },
  },
};
