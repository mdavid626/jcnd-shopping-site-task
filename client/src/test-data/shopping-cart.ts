import { ShoppingCartItem } from '../types/shopping-cart';
import { testProduct1, testProduct2 } from './products';

export const testShoppingCartItem1: ShoppingCartItem = {
  product: testProduct1,
  amount: 3,
};

export const testShoppingCartItem2: ShoppingCartItem = {
  product: testProduct2,
  amount: 36,
};

export const testShoppingCart1: ShoppingCartItem[] = [
  testShoppingCartItem1,
  testShoppingCartItem2,
];
