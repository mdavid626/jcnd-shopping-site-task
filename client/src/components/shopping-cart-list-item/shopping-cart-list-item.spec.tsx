import { cleanup, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import {
  useAddToShoppingCart,
  useRemoveFromShoppingCart,
} from '../../hooks/shopping-car-thooks/shopping-cart-hooks';
import { testShoppingCartItem1 } from '../../test-data/shopping-cart';
import { ShoppingCartItem } from '../../types/shopping-cart';
import ShoppingCartListItem from './shopping-cart-list-item';

jest.mock('../../hooks/shopping-car-thooks/shopping-cart-hooks');
describe('shopping-cart-list-item', () => {
  beforeEach(() => {
    (useAddToShoppingCart as jest.Mock).mockReturnValue(jest.fn());
    (useRemoveFromShoppingCart as jest.Mock).mockReturnValue(jest.fn());
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should render', () => {
    const { asFragment } = render(
      <ShoppingCartListItem shoppingCartItem={testShoppingCartItem1} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useAddToShoppingCart).toHaveBeenCalledWith();
    expect(useRemoveFromShoppingCart).toHaveBeenCalledWith();
  });

  it('should disable add more button when no more in stock', () => {
    const shoppingCartItem: ShoppingCartItem = {
      ...testShoppingCartItem1,
      amount: testShoppingCartItem1.product.inStock,
    };
    render(<ShoppingCartListItem shoppingCartItem={shoppingCartItem} />);
    expect(screen.getByText('add')).toHaveClass(
      'ShoppingCartListItem-button--disabled'
    );
  });

  it('should add more amount when add more button clicked', async () => {
    const addToShoppingCart = jest.fn();
    (useAddToShoppingCart as jest.Mock).mockReturnValue(addToShoppingCart);
    render(<ShoppingCartListItem shoppingCartItem={testShoppingCartItem1} />);
    await userEvent.click(screen.getByText('add'));
    expect(addToShoppingCart).toHaveBeenCalledWith(
      testShoppingCartItem1.product
    );
  });

  it('should remove amount when remove button clicked', async () => {
    const removeFromShoppingCart = jest.fn();
    (useRemoveFromShoppingCart as jest.Mock).mockReturnValue(
      removeFromShoppingCart
    );
    render(<ShoppingCartListItem shoppingCartItem={testShoppingCartItem1} />);
    await userEvent.click(screen.getByText('remove'));
    expect(removeFromShoppingCart).toHaveBeenCalledWith(
      testShoppingCartItem1.product
    );
  });
});
