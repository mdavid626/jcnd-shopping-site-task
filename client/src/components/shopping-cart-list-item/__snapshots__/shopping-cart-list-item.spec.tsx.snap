// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shopping-cart-list-item should render 1`] = `
<DocumentFragment>
  <div
    class="ShoppingCartListItem"
  >
    <div
      class="ShoppingCartListItem-product"
    >
      <div
        class="ShoppingCartListItem-name"
      >
        Artichoke
      </div>
      <div
        class="ShoppingCartListItem-price"
      >
        Price: 46.36 EUR
      </div>
      <div
        class="ShoppingCartListItem-amountColumn"
      >
        <div
          class="ShoppingCartListItem-amount"
        >
          Amount: 3
        </div>
        <div
          class="ShoppingCartListItem-actions"
        >
          <div
            class="ShoppingCartListItem-button"
          >
            add
          </div>
          /
          <div
            class="ShoppingCartListItem-button"
          >
            remove
          </div>
        </div>
      </div>
      <div
        class="ShoppingCartListItem-totalPrice"
      >
        139.08 EUR
      </div>
    </div>
  </div>
</DocumentFragment>
`;
